### PR TITLE
Refactor cluster manager init sequence.

### PR DESCRIPTION
**Pull request by yanjunxiang-google**
This is to address the issue: https://github.com/envoyproxy/envoy/issues/33218

Commit Message:
Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Review by htuch — COMMENTED** [review]
The set of dependencies leading to the need to break the reference circle might benefit from some rethinking longer term, but this seems OK to me. @wbpcode you were interested in reviewing this as well.

File: envoy/upstream/cluster_manager.h [hunk 1]
```diff
@@ -214,6 +214,18 @@ class ClusterManager {
...
+  //
+  // This method *must* be called prior to invoking any other methods on the class and *must* only
+  // be called once. This method should be called immediately after ClusterManagerImpl construction
+  // and from the same thread in which the ClusterManagerImpl was constructed.
+  //
+  // The initialization is separated from the constructor because lots of work, including ADS
+  // initialization, is done in this method. If the contents of this method are invoked during
+  // construction, a derived class cannot override any of the virtual methods and have them invoked
+  // instead, since the base class's methods are used when in a base class constructor.
+  virtual absl::Status init(const envoy::config::bootstrap::v3::Bootstrap& bootstrap) PURE;
```
Comment (wbpcode) [inline envoy/upstream/cluster_manager.h:227]: nit: init() -> initialize()

And I think maybe we could also provide an additional initialized() method to tell if the cluster_manager is initialized?

File: source/server/configuration_impl.cc [hunk 2]
```diff
@@ -128,7 +128,10 @@ absl::Status MainImpl::initialize(const envoy::config::bootstrap::v3::Bootstrap&
   }
 
   ENVOY_LOG(info, "loading {} cluster(s)", bootstrap.static_resources().clusters().size());
+
+  // clusterManagerFromProto() and init() have to be called consecutively.
   cluster_manager_ = cluster_manager_factory.clusterManagerFromProto(bootstrap);
+  RETURN_IF_NOT_OK(cluster_manager_->init(bootstrap));
```
Comment (wbpcode) [inline source/server/configuration_impl.cc:134]: Not sure if this completely resolve our problem. This ensure the server context could always return a valid reference to the cluster manager. But what will happen when the upstream network filter or upstream http filter try to access the actual content of the cluster manager when doing the initialization?

**Review by wbpcode — COMMENTED** [review]
Thanks for your contribution. This seems good to me. (Although I cannot ensure it completely resolve our problem, the status is very complex when doing the initialization. But it at least better then original condition)

**Comment by wbpcode** [comment]
Please check the CI. Then we can try to merge it and see the effect :)

File: envoy/upstream/cluster_manager.h (hunk 1 above)
Comment (yanjunxiang-google) [reply envoy/upstream/cluster_manager.h]: done

File: source/server/configuration_impl.cc (hunk 2 above)
Comment (yanjunxiang-google) [reply source/server/configuration_impl.cc]: Thanks for the comments!

1) When router filter is trying to call processFilters() for upstream filters:
https://github.com/envoyproxy/envoy/blob/63bc9b564b1a76a22a0d029bcac35abeffff2a61/source/common/router/router.cc#L108
it does access context.serverFactoryContext().clusterManager().  IIUC, this is in the much later stage, so the cluster manager instance is already initialized.

2) In the ClusterInfoImpl c'tor, which is called during cluster_manager_->init()
https://github.com/envoyproxy/envoy/blob/63bc9b564b1a76a22a0d029bcac35abeffff2a61/source/common/upstream/upstream_impl.cc#L1376

https://github.com/envoyproxy/envoy/blob/63bc9b564b1a76a22a0d029bcac35abeffff2a61/source/common/upstream/cluster_manager_impl.cc#L905

It's accessing the cluster_manager_, before it is initialized.  This means ClusterInfoImpl c'tor only needs an instance of cluster manager,  and does not require it is initialized.

Thus the change in this PR should be safe.  BTW, there is no actual logical change in this PR.

**Comment by wbpcode** [comment]
LGTM, Thanks for your update. Could you check the CI then we can merge this PR.

**Comment by yanjunxiang-google** [comment]
CI is complaining below. Should be non-related to this change.

Code coverage for source/common/stats is lower than limit of 96.6 (96.5)

**Review by wbpcode — APPROVED** [review]
LGTM. thanks for this contribution. cc @htuch for final pass or merge.

**Comment by yanjunxiang-google** [comment]
Kind ping!

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**